utils/swift_build_support/swift_build_support/products/product.py

```diff
@@ -0,0 +1,32 @@
+# swift_build_support/products/product.py -----------------------*- python -*-
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+
+
+class Product(object):
+    @classmethod
+    def product_name(klass):
+        """product_name() -> str
+
+        The identifier-style name to use for this product.
+        """
+        return klass.__name__.lower()
+
+    @classmethod
+    def get_build_directory_name(klass, host_target):
+        return "{}-{}".format(klass.product_name(),
+                              host_target.name)
+
+    def __init__(self, args, toolchain, source_dir, build_dir):
+        self.args = args
```

Comment:
I investigated this, and I remember why I didn't do it: I want all the products to have the same initializer signature so that the `build-script` can work with the classes generically. I think that is more important than changing the location of the `args` initializer (given that it already leaks out). Instead I think we should just change the Ninja product to stop requiring `args`.